src/client/Microsoft.Identity.Client/Platforms/Android/AndroidBrokerHelper.cs

```diff
@@ -0,0 +1,488 @@
+            string signatureDigest = GetCurrentSignatureForPackage(packageName);
+            if (!string.IsNullOrEmpty(signatureDigest))
+            {
+                return string.Format(CultureInfo.InvariantCulture, "{0}://{1}/{2}", RedirectUriScheme,
+                    packageName.ToLowerInvariant(), signatureDigest);
+            }
+
+            return string.Empty;
+        }
+
+        private string GetCurrentSignatureForPackage(string packageName)
+        {
+            try
+            {
+                PackageInfo info = Application.Context.PackageManager.GetPackageInfo(packageName,
+                    PackageInfoFlags.Signatures);
+                if (info != null && info.Signatures != null && info.Signatures.Count > 0)
+                {
+                    Signature signature = info.Signatures[0];
+                    MessageDigest md = MessageDigest.GetInstance("SHA");
+                    md.Update(signature.ToByteArray());
+                    return Convert.ToBase64String(md.Digest(), Base64FormattingOptions.None);
+                    // Server side needs to register all other tags. ADAL will
+                    // send one of them.
+                }
+            }
+            catch (PackageManager.NameNotFoundException)
+            {
+                _logger.Info("Calling App's package does not exist in PackageManager");
+            }
+            catch (NoSuchAlgorithmException)
+            {
+                _logger.Info("Digest SHA algorithm does not exists");
+            }
+
+            return null;
+        }
+
+        private Intent GetInteractiveBrokerIntent(IDictionary<string, string> brokerPayload, Intent brokerIntent)
+        {
+            ValidateBrokerRedirectURI(brokerPayload);
+            BrokerRequest request = new BrokerRequest
+            {
+                Authority = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.Authority),
+                Scopes = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.Scope),
+                RedirectUri = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.RedirectUri),
+                ClientId = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.ClientId),
+                ClientAppName = Application.Context.PackageName,
+                ClientAppVersion = Application.Context.PackageManager.GetPackageInfo(Application.Context.PackageName, PackageInfoFlags.MatchAll).VersionName,
+                ClientVersion = "4.4.0",
+                CorrelationId = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.CorrelationId),
+                Prompt = "NONE"
+            };
+
+            brokerIntent.PutExtra(BrokerConstants.BrokerRequestV2, JsonHelper.SerializeToJson(request));
+
+            return brokerIntent;
+        }
+
+        private Bundle GetSilentBrokerBundle(IDictionary<string, string> brokerPayload)
+        {
+            Bundle bundle = new Bundle();
+
+            BrokerRequest request = new BrokerRequest
```

Comment:
this is the same as above...i thought silent didn't have a redirecturi? can this be moved to one method?

Reply:
Silent is needed for broker because it uses the redirect URI as a key to lookup accounts in the account manager i believe.